h2o-genmodel/src/main/java/hex/genmodel/algos/tree/ScoreTree2.java

```diff
@@ -1,10 +1,477 @@
+
+      SharedTreeNode newNode = sg.makeRightChildNode(node);
+      newNode.setWeight(auxInfo.weightR);
+      newNode.setNodeNumber(auxInfo.nidR);
+      newNode.setPredValue(auxInfo.predR);
+      newNode.setSquaredError(auxInfo.sqErrR);
+      if ((lmask2 & 16) != 0) {
+        float leafValue = ab2.get4f();
+        newNode.setPredValue(leafValue);
+        auxInfo.predR = leafValue;
+      } else {
+        computeTreeGraph(sg, newNode, tree, ab2, auxMap, names, domains);
+      }
+    }
+
+    // go LEFT
+    {
+      ByteBufferWrapper ab2 = new ByteBufferWrapper(tree);
+      ab2.skip(ab.position());
+
+      if (lmask <= 3)
+        ab2.skip(lmask + 1);
+
+      SharedTreeNode newNode = sg.makeLeftChildNode(node);
+      newNode.setWeight(auxInfo.weightL);
+      newNode.setNodeNumber(auxInfo.nidL);
+      newNode.setPredValue(auxInfo.predL);
+      newNode.setSquaredError(auxInfo.sqErrL);
+      if ((lmask & 16) != 0) {
+        float leafValue = ab2.get4f();
+        newNode.setPredValue(leafValue);
+        auxInfo.predL = leafValue;
+      } else {
+        computeTreeGraph(sg, newNode, tree, ab2, auxMap, names, domains);
+      }
+    }
+    if (node.getNodeNumber() == 0) {
+      float p = (float) (((double) auxInfo.predL * (double) auxInfo.weightL + (double) auxInfo.predR * (double) auxInfo.weightR) / ((double) auxInfo.weightL + (double) auxInfo.weightR));
+      if (Math.abs(p) < 1e-7) p = 0;
+      node.setPredValue(p);
+      node.setSquaredError(auxInfo.sqErrR + auxInfo.sqErrL);
+      node.setWeight(auxInfo.weightL + auxInfo.weightR);
+    }
+    checkConsistency(auxInfo, node);
+  }
+
+  private static void checkConsistency(AuxInfo auxInfo, SharedTreeNode node) {
+    boolean ok = true;
+    ok &= (auxInfo.nid == node.getNodeNumber());
+    double sum = 0;
+    if (node.leftChild != null) {
+      ok &= (auxInfo.nidL == node.leftChild.getNodeNumber());
+      ok &= (auxInfo.weightL == node.leftChild.getWeight());
+      ok &= (auxInfo.predL == node.leftChild.predValue);
+      ok &= (auxInfo.sqErrL == node.leftChild.squaredError);
+      sum += node.leftChild.getWeight();
+    }
+    if (node.rightChild != null) {
+      ok &= (auxInfo.nidR == node.rightChild.getNodeNumber());
+      ok &= (auxInfo.weightR == node.rightChild.getWeight());
+      ok &= (auxInfo.predR == node.rightChild.predValue);
+      ok &= (auxInfo.sqErrR == node.rightChild.squaredError);
+      sum += node.rightChild.getWeight();
+    }
+    if (node.parent != null) {
+      ok &= (auxInfo.pid == node.parent.getNodeNumber());
+      ok &= (Math.abs(node.getWeight() - sum) < 1e-5 * (node.getWeight() + sum));
+    }
+    if (!ok) {
+      System.out.println("\nTree inconsistency found:");
+      node.print();
+      node.leftChild.print();
+      node.rightChild.print();
+      System.out.println(auxInfo.toString());
+    }
+  }
+
+  private static HashMap<Integer, AuxInfo> readAuxInfos(ByteBufferWrapper abAux) {
```

Comment:
Why are these just in ScoreTree2?

Reply:
here are indications that led me to this:
- there is only one implementation of `computeTreeGraph`, and based on the fact that its implementation uses the infamous `bs.fill3` method, it is clearly V2 only (V0 and V1 both use `bs.fill3_1` instead)
- this V2-only `computeTreeGraph` method is the one called from `readAuxInfos` which makes it V2-only as well, though almost the same code can be used for V0 and V1

I am almost sure that there was no way to call `readAuxInfos` on V1 (or V0 if they had auxinfos) model, without risking error during tree traversal. The error happens when there is a model with trees using domain with more than 32 levels (long bitset) - that's where the storage format differs.

Reply:
The question also shows that, having all these three versions in single class is confusing, and makes the lack of functionality for older version invisible. Or at least, very hard to see.

If we decide to add it, another PR should do that IMHO.